Add JS error boundaries

a-danae opened the pull request:
Fixes #1147 

#### Changes proposed in this Pull Request

Add error boundaries to prevent the entire app from unmounting on error.

As mentioned in the issue, a JS error causes the entire app to unmount (as witnessed in #1145). The goal of this PR is to prevent the entire app from unmounting when a JS error is triggered while giving a useful message to users, and maybe logging/tracking the error.

#### Testing instructions

I think the easiest way to test this is by updating the code to throw errors in different places.

Here's an example of how a screen looks like. Before:
![image](https://user-images.githubusercontent.com/41606954/136582512-e0d31bfb-2b72-4c93-80aa-e01d9354f8f0.png)

With the added boundary:
![image](https://user-images.githubusercontent.com/41606954/136581771-8e8254ae-539a-4614-a429-0af1d26cbd2d.png)


-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

a-danae commented on client/components/error-boundary/index.js:
```diff
@@ -0,0 +1,45 @@
+
+		return (
+			<Notice status="error" isDismissible={ false }>
+				{ __(
+					'There was an error rendering the component, please try again. Contact support for assistance if the problem persists.',
+					'woocommerce-payments'
+				) }
+				<br />
+				{ this.state.error.toString() }
```
We could probably update the error message to make it more useful for users. Glad to adjust it.

a-danae commented on client/connect-account-page/index.js:
```diff
@@ -11,6 +11,7 @@ import { render, useState, useEffect } from '@wordpress/element';
  */
 import OnboardingLocationCheckModal from './modal';
 import Page from 'components/page';
+import ErrorBoundary from 'components/error-boundary';
```
I'm wrapping the content of all pages within ErrorBoundary. We still have to define how we'd like to track errors, but it might be better to wrap the page's contents directly in the [Page component](https://github.com/Automattic/woocommerce-payments/blob/develop/client/components/page/index.tsx#L43) instead of doing it on each page.

I'm doing it this way for now in case each page would trigger different actions. But if this won't be the case, I think having it within the Page component instead would be neater. Glad to learn if some other approach would be better, though!

dechov commented on client/components/error-boundary/index.js:
```diff
@@ -0,0 +1,45 @@
+		return (
+			<Notice status="error" isDismissible={ false }>
+				{ __(
+					'There was an error rendering the component, please try again. Contact support for assistance if the problem persists.',
```
I wouldn't mention "the component" in the user-facing error since it's kind of technical – maybe "this view"? Also "please try again" isn't necessarily actionable from the user's perspective, though I don't really have an alternative in mind.

a-danae replied:
Makes sense, thanks! Copy updated. I've removed the "try again" part. I think it reads okay and doesn't give a possibly non-actionable instruction. Let me know if we better adjust it further!

a-danae commented:
Removing you, @dechov, after our chat. I've adjusted the notice in [6a4686e](https://github.com/Automattic/woocommerce-payments/pull/3034/commits/6a4686e10605d6edcf3320b15368032b27bf69b3) because of the alignment detail you mentioned. It should now look as in the screenshot below. Thanks for checking this out!

![image](https://user-images.githubusercontent.com/41606954/145129661-b2ae3812-3b10-4b97-8eef-4ac23e3cb111.png)

@leonardola requesting you another review here because of this change. We were pinging other teams for reviewing the PRs back when I opened this one, but that's up to you now that we've changed that flow 🙂

leonardola commented:
LGTM. Nice job. You can go ahead and merge it

dechov commented:
Haven't tested but code changes LGTM – thanks for addressing this!

(Note before merging that changelog version should be adjusted again.)

a-danae commented:
Cool! Changelog entry updated and proceeding to merge.

Thank you both for reviewing this!
